sql/src/main/java/io/deephaven/sql/RelNodeVisitorAdapter.java

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.sql;
+
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelShuttle;
+import org.apache.calcite.rel.core.TableFunctionScan;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.logical.LogicalAggregate;
+import org.apache.calcite.rel.logical.LogicalCalc;
+import org.apache.calcite.rel.logical.LogicalCorrelate;
+import org.apache.calcite.rel.logical.LogicalExchange;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rel.logical.LogicalIntersect;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rel.logical.LogicalMatch;
+import org.apache.calcite.rel.logical.LogicalMinus;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.logical.LogicalSort;
+import org.apache.calcite.rel.logical.LogicalTableModify;
+import org.apache.calcite.rel.logical.LogicalUnion;
+import org.apache.calcite.rel.logical.LogicalValues;
+
+import java.util.Objects;
+
+final class RelNodeVisitorAdapter<T> implements RelShuttle {
+
+    private final RelNodeVisitor<T> visitor;
+    private T out;
+
+    public RelNodeVisitorAdapter(RelNodeVisitor<T> visitor) {
+        this.visitor = Objects.requireNonNull(visitor);
+    }
+
+    public T out() {
+        return Objects.requireNonNull(out);
+    }
+
+    @Override
+    public RelNode visit(TableScan scan) {
+        out = visitor.visit(scan);
+        return scan;
+    }
+
+    @Override
+    public RelNode visit(TableFunctionScan scan) {
+        // SQLTODO(custom-sources)
+        //
+        // It would be good to be able to source tables from other places, besides just ones in the global scope. For
+        // example:
+        // SELECT * FROM parquet('/data/my-dataset.parquet')
+        // SELECT * FROM csv('/my/test.csv')
+        // SELECT * FROM uri('dh://server/scope/table_name')
+        // SELECT * FROM uri('csv:///path/to/the.csv')
+        //
+        // Potentially related to design decisions around SQLTODO(catalog-reader-implementation)
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public RelNode visit(LogicalValues values) {
+        out = visitor.visit(values);
+        return values;
+    }
+
+    @Override
+    public RelNode visit(LogicalFilter filter) {
+        out = visitor.visit(filter);
+        return filter;
+    }
+
+    @Override
```

Comment:
Maybe this is updateView?

Reply:
A LogicalCalc is:

> A relational expression which computes project expressions and also filters.

I think it only comes into play when Calcite optimizations are turned on.